consensus/src/lib.rs

```diff
@@ -160,29 +162,27 @@ impl<N: GossipNetwork, S: Storage> Consensus<N, S> {
         round_zero_timestamp: Timestamp,
         this_node_key: Option<PrivateKey>,
     ) -> Result<Self, Error> {
-        let state = state_storage.read_file(STATE_FILE_NAME).await?;
-        let state: State = serde_json::from_str(&state)?;
-        // TODO: check if `this_node_key` is in the validator set. If not, error.
-        if block_header != state.block_header {
-            let height_info = generate_height_info(
-                &block_header,
-                consensus_parameters,
-                round_zero_timestamp,
-                this_node_key.clone(),
-            )?;
+        // Do a pre-calculation for prettier code.
+        let new_state = Self::construct_new_state(
+            &block_header,
+            consensus_parameters,
+            round_zero_timestamp,
+            this_node_key.clone().unwrap(),
+        )?;
+        let state = if let Ok(raw_state) = state_storage.read_file(STATE_FILE_NAME).await {
+            let state: State = serde_json::from_str(&raw_state)?;
+            if block_header != state.block_header {
+                dms.clear(generate_dms_key(&block_header)).await?;
+                commit_state(&mut state_storage, &new_state).await?;
+                new_state
+            } else {
+                state
+            }
+        } else {
             dms.clear(generate_dms_key(&block_header)).await?;
-            let state = State {
-                vetomint: Vetomint::new(height_info),
-                block_header,
-                updated_messages: BTreeSet::new(),
-                verified_block_hashes: vec![],
-                vetoed_block_hashes: vec![],
-                finalized: false,
-            };
-            state_storage
-                .add_or_overwrite_file(STATE_FILE_NAME, serde_json::to_string(&state).unwrap())
-                .await?;
-        }
+            commit_state(&mut state_storage, &new_state).await?;
```

Comment:
You have to clear the `state_storage` as well.

Reply:
fixed